### ISSUE TITLE
automatically ignore monorepo workspaces

### ISSUE THREAD

**Issue by dylang** [renovateapp/renovate#1398]
This should never happen for yarn monorepo workspaces, because yarn automatically symlinks these dependencies instead of loading them from the registry:

>  INFO: Failed to look up dependency. This will block *all* dependencies from being updated due to presence of lock file. (repository=org/monorepo, packageFile=package.json, depType=devDependencies, dependency=@scope/package-in-monorepo)

I think the expected behavior is to ignore all dependencies/devDeps that are in the repo as a workspace, because there's no reason to upgrade them as they are symlinked in the repo.

**Comment by rarkins** [comment]
@dylang this functionality as you expect is in the codebase and is/was working with quite a lot of monorepos the last I checked it. Can you work out a way to reproduce it in a public way, or give more details?

BTW this feature is essentially just an "automatic ignore" one where it scans the directories declared inside `workspaces`, then extracts the `name` field, and adds them all to the implicit ignore list for all other packages within the same workspace. If you are in a hurry to get it working, you can use `ignoreDeps` or `packageRules` to manually ignore such packages.

**Comment by rarkins** [comment]
> I think the expected behavior is to ignore all dependencies/devDeps that are in the repo as a workspace

Not exactly. It ignores all package names that are listed in the "name" field of `package.json` files within the workspace, but it only ignores them *within that workspace*. I have a user where one repository has a combination of workspace and non-workspace and needs packages outside the workspace to still have normal behaviour and perform npm lookups. Renovate's logic is meant to even support multiple workspaces independently within the same repository.

**Comment by dylang** [comment]
> within that workspace

Do you mean "within that set of workspace"?

Not the same workspace, that would mean depending on itself.

**Comment by rarkins** [comment]
I may be lost in terminology, so I'll give an example. You could have a setup where you have "app1/package.json" containing a "yarn.lock" and a "workspaces" definition with many subpackages inside that, and you could also have a "app2/package.json" with the same, that operates independently. In this case the implicit "ignore" is isolated to each app1 and does not cross-ignore. e.g. if package "foobarbaz" is defined within "app1/packages/foo-bar-baz/package.json" then it would be ignored by all app1 packages but not by all app2 ones.

**Comment by dylang** [comment]
We have something like this:

```
--packages
   -- package1
       -- package.json
   -- package2
      -- package.json
package.json
yarn.lock
```

Workspaces are `packages/*`. The `yarn.lock` in the root is for the entire repo.

The repo root is not in the workspace because `yarn`'s docs say workspaces can't be nested.

However, that `package.json` in the repo root has dependencies that are workspaces. This is how we support things like `yarn start` and `yarn test` from the repo root. 

I think you are saying that, because the repo root is not explicitly listed as a workspace, those dependencies that are workspaces must be explicitly ignored in the configuration.

**Comment by rarkins** [comment]
No, in that case I agree that if any of those sub-packages are listed in the root's package.json then they should be ignored as well. I will check tomorrow in case that is not happening.

**Comment by rarkins** [comment]
@dylang I think that you're right - the implicit ignore for subpackages was not being done *in the root package.json* for the workspaces. I have added an extra check for that and released in renovate@10.51.2 - could you retry?
